### PR TITLE
[Hotfix][Pixtral] Fix multiple images bugs

### LINKED ISSUE #8382: [Bug]: Pixtral fails when limit_mm_per_prompt not set
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
Your output of `python collect_env.py` here
```

</details>

### 🐛 Describe the bug

The below command does not work
```
CUDA_VISIBLE_DEVICES=3 vllm serve mistralai/Pixtral-12B-2409 --port 21010 --max_num_batched_tokens 16384 --trust-remote-code --gpu-memory-utilization 0.50 --tokenizer_mode mistral
```

It leads to this error: 
```
Traceback (most recent call last):
  File "/home/lmsys/miniconda3/envs/vllm-source/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/home/lmsys/miniconda3/envs/vllm-source/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/lmsys/vllm/vllm/entrypoints/openai/rpc/server.py", line 236, in run_rpc_server
    server = AsyncEngineRPCServer(async_engine_args, usage_context, rpc_path)
  File "/home/lmsys/vllm/vllm/entrypoints/openai/rpc/server.py", line 34, in __init__
    self.engine = AsyncLLMEngine.from_engine_args(
  File "/home/lmsys/vllm/vllm/engine/async_llm_engine.py", line 735, in from_engine_args
    engine = cls(
  File "/home/lmsys/vllm/vllm/engine/async_llm_engine.py", line 615, in __init__
    self.engine = self._init_engine(*args, **kwargs)
  File "/home/lmsys/vllm/vllm/engine/async_llm_engine.py", line 835, in _init_engine
    return engine_class(*args, **kwargs)
  File "/home/lmsys/vllm/vllm/engine/async_llm_engine.py", line 262, in __init__
    super().__init__(*args, **kwargs)
  File "/home/lmsys/vllm/vllm/engine/llm_engine.py", line 338, in __init__
    self._initialize_kv_caches()
  File "/home/lmsys/vllm/vllm/engine/llm_engine.py", line 467, in _initialize_kv_caches
    self.model_executor.determine_num_available_blocks())
  File "/home/lmsys/vllm/vllm/executor/gpu_executor.py", line 114, in determine_num_available_blocks
    return self.driver_worker.determine_num_available_blocks()
  File "/home/lmsys/miniconda3/envs/vllm-source/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/home/lmsys/vllm/vllm/worker/worker.py", line 223, in determine_num_available_blocks
    self.model_runner.profile_run()
  File "/home/lmsys/miniconda3/envs/vllm-source/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/home/lmsys/vllm/vllm/worker/model_runner.py", line 1216, in profile_run
    self.execute_model(model_input, kv_caches, intermediate_tensors)
  File "/home/lmsys/miniconda3/envs/vllm-source/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/home/lmsys/vllm/vllm/worker/model_runner.py", line 1543, in execute_model
    hidden_or_intermediate_states = model_executable(
  File "/home/lmsys/miniconda3/envs/vllm-source/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/lmsys/miniconda3/envs/vllm-source/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/lmsys/vllm/vllm/model_executor/models/pixtral.py", line 178, in forward
    inputs_embeds = merge_multimodal_embeddings(
  File "/home/lmsys/vllm/vllm/model_executor/models/pixtral.py", line 117, in merge_multimodal_embeddings
    assert (seq_len == N_txt +
AssertionError: seq_len 16640 should be equal to N_txt + N_img (256, 4096, 16384)
```

But the below works (following huggingface):
```
CUDA_VISIBLE_DEVICES=3 vllm serve mistralai/Pixtral-12B-2409 --port 21010 --max_num_batched_tokens 16384 --max-model-len 8192 --trust-remote-code --gpu-memory-utilization 0.50 --tokenizer_mode mistral --limit_mm_per_prompt 'image=4'
```

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### LINKED ISSUE #8411: [Bug]: Pixtral inference not working correctly with LLMEngine/AsyncEngine
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
Collecting environment information...
PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jul 29 2024, 16:56:48) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-119-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100 80GB PCIe
Nvidia driver version: 535.183.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               12
On-line CPU(s) list:                  0-11
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Gold 6342 CPU @ 2.80GHz
CPU family:                           6
Model:                                106
Thread(s) per core:                   1
Core(s) per socket:                   12
Socket(s):                            1
Stepping:                             6
BogoMIPS:                             5600.12
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush acpi mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single intel_ppin ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves umip pku ospke gfni vaes vpclmulqdq rdpid md_clear flush_l1d arch_capabilities
Hypervisor vendor:                    Xen
Virtualization type:                  full
L1d cache:                            576 KiB (12 instances)
L1i cache:                            384 KiB (12 instances)
L2 cache:                             15 MiB (12 instances)
L3 cache:                             432 MiB (12 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-11
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-ml-py==11.525.150
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] nvidia-nvjitlink-cu12==12.6.68
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] pyzmq==26.2.0
[pip3] torch==2.4.0
[pip3] torchvision==0.19.0
[pip3] transformers==4.44.2
[pip3] triton==3.0.0
[conda] Could not collect
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.6.1@3fd2b0d21cd9ec78de410fdf8aa1de840e9ad77a
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
[4mGPU0	CPU Affinity	NUMA Affinity	GPU NUMA ID[0m
GPU0	 X 	0-11	0		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks
```

</details>

### Model Input Dumps

_No response_

### 🐛 Describe the bug

This code snippets does not work

```python
import PIL.Image
import uuid
from vllm import EngineArgs, LLMEngine
from vllm import SamplingParams, TextPrompt
from vllm.multimodal import MultiModalDataBuiltins

MODEL_ID = "mistral-community/pixtral-12b-240910"
ENGINE_ARGS = EngineArgs(
    model=MODEL_ID,
    tokenizer_mode="mistral",
    enable_prefix_caching=True,
    limit_mm_per_prompt=dict(image=4),
    max_num_batched_tokens=16384,
)
SAMPLING_PARAM = SamplingParams()

engine = LLMEngine.from_engine_args(ENGINE_ARGS)
prompt = "describe the image"

engine_inputs = TextPrompt(prompt=prompt)

image = PIL.Image.open("demo.jpg")
mm_data = MultiModalDataBuiltins(image=[image])
engine_inputs["multi_modal_data"] = mm_data

engine.add_request(uuid.uuid4().hex, engine_inputs, SAMPLING_PARAM)

while True:
    out = engine.step()
    for request_output in request_outputs:
        if request_output.finished:
            print(request_output)
    if not engine.has_unfinished_requests():
        break
```

This will give an output message like:

```
 File "/workspace/codes/example/vllm/pixtral-12b/venv/lib/python3.10/site-packages/vllm/model_executor/models/pixtral.py", line 117, in merge_multimodal_embeddings
    assert (seq_len == N_txt +
AssertionError: seq_len 7 should be equal to N_txt + N_img (7, 1200, 0)
```

It seems that I need to manually padding the input token ids with image_token_ids like this:

```python
tokenizer = engine.get_tokenizer()
token_ids = tokenizer(prompt).input_ids
image_token_id = 10
token_ids = [image_token_id] * image_token_num + token_ids

engine_inputs = TokensPrompt(prompt_token_ids=token_ids)

mm_data = MultiModalDataBuiltins(image=[image])
engine_inputs["multi_modal_data"] = mm_data
```

To make it work.

`AsyncLLMEngine` also see the same limitation and I need similar modification to make it works like: https://github.com/bentoml/BentoVLLM/pull/71/files#diff-357f77bce00e63217bb5ec382293bca653276e58af9bdfb6e7c50ca9487e27aeR84-R94

Is this behavior intended or a bug?

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by patrickvonplaten**
FILL IN THE PR DESCRIPTION HERE

FIX #8382
FIX #8411

This PR fixes a couple bugs that arrive from using chunked pre-filling and previously incorrect image processing.

This PR makes sure that all images are pre-processed correctly and adds a bunch of aggressive tests.

### CODE REVIEW

File: vllm/model_executor/models/pixtral.py [hunk 1]
```diff
@@ -115,8 +122,9 @@ def merge_multimodal_embeddings(input_ids: torch.Tensor,
     assert (D_txt == D_img), (f"Text features dim {D_txt} should be equal "
                               "to image features dim {D_img}")
     assert (seq_len == N_txt +
-            N_img), (f"seq_len {seq_len} should be equal to N_txt + N_img "
-                     f"{(N_txt, N_img, image_locations.sum().item())}")
+            N_img), (f"seq_len should be equal to N_txt + N_img, "
+                     f" but got {seq_len=} != {N_txt=} + {N_img=}, "
+                     f" for {image_locations.sum().item()} img locs")
```
Comment (DarkLight1337) [inline vllm/model_executor/models/pixtral.py:127]: By the way, is it intended that you use a separate function than the built in one in `models/utils.py`?

**Comment by DarkLight1337** [comment]
Would be great if you could add a test to avoid similar regressions!

File: vllm/model_executor/models/pixtral.py (hunk 1 above)
Comment (patrickvonplaten) [reply vllm/model_executor/models/pixtral.py:127]: Nice catch - I think we can safely delete our custom merge function here

File: tests/models/test_pixtral.py [hunk 2]
```diff
@@ -2,63 +2,156 @@
 
 Run `pytest tests/models/test_mistral.py`.
 """
+import uuid
```
Comment (patrickvonplaten) [inline tests/models/test_pixtral.py:5]: Add many more aggressive tests

**Comment by DarkLight1337** [comment]
The tests fail to pass when I run them locally.

**Comment by patrickvonplaten** [comment]
> The tests fail to pass when I run them locally.

Hmm I see what device to you run them on? Tested on H100 - for me they are passing

**Comment by patrickvonplaten** [comment]
> > The tests fail to pass when I run them locally.
> 
> Hmm I see what device to you run them on? Tested on H100 - for me they are passing

Guess it's going to be quite difficult to get exactly the same results here across different devices and given that flash attention is not that deterministic of an operation

**Comment by DarkLight1337** [comment]
> > The tests fail to pass when I run them locally.
> 
> Hmm I see what device to you run them on? Tested on H100 - for me they are passing

I'm running the test on a single L40. (I can't fit `max_model_len=65536` so can only run the smaller tests)

**Comment by patrickvonplaten** [comment]
> > > The tests fail to pass when I run them locally.
> > 
> > 
> > Hmm I see what device to you run them on? Tested on H100 - for me they are passing
> 
> I'm running the test on a single L40. (I can't fit `max_model_len=65536` so can only run the smaller tests)

Ok yes that doesn't surprise me. From your test failure it seems like the first test case passes but then the longer / more complicated tests fails. We could add device-dependent expected values? Not sure. Wdyt?

**Comment by DarkLight1337** [comment]
> > > The tests fail to pass when I run them locally.
> > 
> > 
> > Hmm I see what device to you run them on? Tested on H100 - for me they are passing
> 
> I'm running the test on a single L40. (I can't fit `max_model_len=65536` so can only run the smaller tests)

I think the output is still reasonable. Since the goal of this PR is to avoid crashing the model rather than having perfect output, we can reduce the number of tokens to check for now (at least until we have a HF version to test against).

**Comment by patrickvonplaten** [comment]
> > > > The tests fail to pass when I run them locally.
> > > 
> > > 
> > > Hmm I see what device to you run them on? Tested on H100 - for me they are passing
> > 
> > 
> > I'm running the test on a single L40. (I can't fit `max_model_len=65536` so can only run the smaller tests)
> 
> I think the output is still reasonable. Since the goal of this PR is to avoid crashing the model rather than having perfect output, we can reduce the number of tokens to match for now (at least until we have a HF version to test against).

I'll comment out the more extreme tests and for now will just them locally when needed

**Comment by patrickvonplaten** [comment]
Wrapped the more extreme tests in a `is_h100` wrapper - does that work?

**Comment by DarkLight1337** [comment]
Imo that's too device-specific. If you're able to extract the logprobs information, it would be better to check against the golden output via `check_logprobs_close`.

**Comment by DarkLight1337** [comment]
As for the OOM issue, we can address that via TP in another PR.

**Comment by patrickvonplaten** [comment]
> Imo that's too device-specific. If you're able to extract the logprobs information, it would be better to check against the golden output via `check_logprobs_close`.

Hmm what is the golden output for you then? Think vLLM should represent the official implementation here

**Comment by DarkLight1337** [comment]
You can use your H100 output as the golden one. Checking the logprobs is less strict so the test should still pass on other devices.

**Comment by patrickvonplaten** [comment]
> You can use your H100 output as the golden one. Checking the logprobs is less strict so the test should still pass on other devices.

Do you check that logprobs are within a range? If the output is different for temperature=0.0, the logprobs also quite certainly won't match no? In my experience it's quite difficult to get outputs to exactly match over different devices except for small inputs. I can try to extract the logprobs, but they also won't match between L40 and H100 I'm afraid

**Comment by DarkLight1337** [comment]
> > You can use your H100 output as the golden one. Checking the logprobs is less strict so the test should still pass on other devices.
> 
> Do you check that logprobs are within a range? If the output is different for temperature=0.0, the logprobs also quite certainly won't match no? In my experience it's quite difficult to get outputs to exactly match over different devices except for small inputs. I can try to extract the logprobs, but they also won't match between L40 and H100 I'm afraid

We just check that for each token outputted by vLLM, the selected token is within the top-k logprobs of the reference (golden) output.

**Comment by patrickvonplaten** [comment]
> > > You can use your H100 output as the golden one. Checking the logprobs is less strict so the test should still pass on other devices.
> > 
> > 
> > Do you check that logprobs are within a range? If the output is different for temperature=0.0, the logprobs also quite certainly won't match no? In my experience it's quite difficult to get outputs to exactly match over different devices except for small inputs. I can try to extract the logprobs, but they also won't match between L40 and H100 I'm afraid
> 
> We just check that for each token outputted by vLLM, the selected token is within the top-k logprobs of the reference (golden) output. You can use the `check_logprobs_close` utility function for this.

Gotcha!

**Comment by patrickvonplaten** [comment]
> > > > You can use your H100 output as the golden one. Checking the logprobs is less strict so the test should still pass on other devices.
> > > 
> > > 
> > > Do you check that logprobs are within a range? If the output is different for temperature=0.0, the logprobs also quite certainly won't match no? In my experience it's quite difficult to get outputs to exactly match over different devices except for small inputs. I can try to extract the logprobs, but they also won't match between L40 and H100 I'm afraid
> > 
> > 
> > We just check that for each token outputted by vLLM, the selected token is within the top-k logprobs of the reference (golden) output. You can use the `check_logprobs_close` utility function for this.
> 
> Gotcha!

Actually sorry even this won't work because errors accumulate and then context changes and the ouput results is very different. E.g. for second example, I can get:

**Gold output**:
```
"1. A black dog with floppy ears sits attentively on a wooden surface.\n2. A vast mountain range with rugged peaks stretches under a cloudy sky.",
```

and

**L40**:
```
"1. A black dog with floppy ears sits attentively on a wooden surface.\n2. A vast mountain range stretches across the horizon under a cloudy sky."
```

Here the first different word is `"with"` vs. `"stretches"`. "stretches" will be in the topk range of logprobs of "with" but the next words will not be. Currently each of the two tests run a simple test on every device where results match. Just the more difficult tests are only run on H100 so on L40 still two tests are run.

File: tests/models/test_pixtral.py [hunk 3]
```diff
@@ -2,63 +2,177 @@
...
+
+MAX_MODEL_LEN = [8192, 65536] if is_h100_gpu() else [8192]
 
 
 @pytest.mark.skip(
     reason=
     "Model is too big, test passed on A100 locally but will OOM on CI machine."
 )
 @pytest.mark.parametrize("model", MODELS)
+@pytest.mark.parametrize("max_model_len", [8192, 65536])
```
Comment (patrickvonplaten) [inline tests/models/test_pixtral.py:107]:
```suggestion
@pytest.mark.parametrize("max_model_len", MAX_MODEL_LEN)
```

**Review by ywang96 — APPROVED** [review]
I've run the new test on H100 and it all passed for me too, so I'm giving this a green light. As for the refactor work on the test so that it can run on the L40 machines on our CI, let's do that in a later PR given our timeline for the patch release.

Thanks for fixing!

**Comment by patrickvonplaten** [comment]
Just added the logprobs tests as explained by @DarkLight1337 - think it indeed makes more sense! Would be great if it passes also on a L40. Thanks for the reviews!

**Pull request merged**